src/m365/pa/commands/app/app-permission-list.ts

```diff
@@ -0,0 +1,131 @@
+import { Logger } from '../../../../cli/Logger';
+import GlobalOptions from '../../../../GlobalOptions';
+import { formatting } from '../../../../utils/formatting';
+import { odata } from '../../../../utils/odata';
+import { validation } from '../../../../utils/validation';
+import PowerAppsCommand from '../../../base/PowerAppsCommand';
+import commands from '../../commands';
+
+interface CommandArgs {
+  options: Options;
+}
+
+interface Options extends GlobalOptions {
+  appName: string;
+  asAdmin: boolean;
+  environment?: string;
+  roleName?: string;
+}
+
+class PaAppPermissionListCommand extends PowerAppsCommand {
+  private allowedRoleNames = ['Owner', 'CanEdit', 'CanView'];
+
+  public get name(): string {
+    return commands.APP_PERMISSION_LIST;
+  }
+
+  public get description(): string {
+    return 'Lists all permissions of a Power Apps app';
+  }
+
+  public defaultProperties(): string[] | undefined {
+    return ['roleName', 'principalId', 'principalType'];
+  }
+
+  constructor() {
+    super();
+
+    this.#initTelemetry();
+    this.#initOptions();
+    this.#initValidators();
+    this.#initOptionSets();
+  }
+
+  #initTelemetry(): void {
+    this.telemetry.push((args: CommandArgs) => {
+      Object.assign(this.telemetryProperties, {
+        asAdmin: !!args.options.asAdmin,
+        environmentName: typeof args.options.environment !== 'undefined',
+        roleName: typeof args.options.roleName !== 'undefined'
+      });
+    });
+  }
+
+  #initOptions(): void {
+    this.options.unshift(
+      {
+        option: '--appName <appName>'
+      },
+      {
+        option: '--asAdmin'
+      },
+      {
+        option: '-e, --environmentName [environmentName]'
+      },
+      {
+        option: '--roleName [roleName]',
+        autocomplete: this.allowedRoleNames
+      }
+    );
+  }
+
+  #initValidators(): void {
+    this.validators.push(
+      async (args: CommandArgs) => {
+        if (!validation.isValidGuid(args.options.appName)) {
+          return `${args.options.appName} is not a valid GUID.`;
+        }
+
+        if (args.options.roleName && !this.allowedRoleNames.includes(args.options.roleName)) {
+          return `${args.options.roleName} is not a valid roleName. Allowed values are ${this.allowedRoleNames.join('|')}`;
+        }
+
+        return true;
+      }
+    );
+  }
+
+  #initOptionSets(): void {
+    this.optionSets.push(
+      {
+        options: ['environmentName'],
+        runsWhen: (args) => args.options.asAdmin
```

Comment:
This works indeed, however, the error message looks a bit odd.
`Specify one of the following options: environmentName`

Let's just add our own validation with a better message. We should also check that `environmentName` is **not** passed when `asAdmin` is **not** specified.